### PR TITLE
Feature/ua 1163

### PR DESCRIPTION
**Pull request by vcward**
Added a `Series.load_from_eia()` definition to pull data from the [EIA API](https://www.eia.gov/opendata/qb.php). I added the API key as an environment variable to prod (API_KEY_EIA).
Example [series](https://www.eia.gov/opendata/qb.php?category=430&sdid=ELEC.CONS_TOT.PEL-HI-99.A)  `Series.load_from_eia('ELEC.CONS_TOT.PEL-HI-99.A')`

### CODE REVIEW

File: app/lib/data_html_parser.rb [hunk 1]
```diff
@@ -88,6 +88,30 @@ def get_clustermapping_series(dataset, parameters)
     new_data
   end
 
+  def get_eia_series(parameter)
+    api_key = ENV['API_KEY_EIA']
+    raise 'No API key defined for EIA' unless api_key
+    @url = "http://api.eia.gov/series/?series_id=#{parameter}&api_key=#{api_key}"
```
Comment (kotodharma) [inline app/lib/data_html_parser.rb]: Although there is a redirect in place, might be better to have `https` directly in the code :=)

File: app/lib/data_html_parser.rb [hunk 2]
```diff
@@ -145,6 +169,7 @@ def get_freq(other_string)
   def get_date(year_string, other_string)
     return Date.new(year_string.to_i) if other_string == 'M13'
     return Date.new(year_string.to_i, other_string[1..2].to_i) unless %w(M01 M02 M03 M04 M05 M06 M07 M08 M09 M10 M11 M12).index(other_string).nil?
+    return Date.new(year_string.to_i, other_string.to_i) unless %w(01 02 03 04 05 06 07 08 09 10 11 12).index(other_string).nil?
```
Comment (kotodharma) [inline app/lib/data_html_parser.rb]: I have no reason to think there's a problem here, but just wanted to register a note of caution: please make extra sure that by inserting this line in this position, you're not accidentally breaking some behavior implemented in the following lines. I really hate this coding style (not your fault :=) but it's what we've got in so many places in udaman and changing it here, now, is definitely not worth the effort :=/

Comment (vcward) [reply app/lib/data_html_parser.rb]: Got it, will try to check through some other series, hopefully that didn't break anything.

File: app/lib/data_html_parser.rb [hunk 3]
```diff
@@ -165,18 +165,29 @@ def get_freq(other_string)
...
+      Date.new(year_string.to_i, 4)
+    # (Quarterly/Semi-Annual) Observations that should evaluate to July of year_string
+    when /^(S02|Q(3|03))\b/
+      Date.new(year_string.to_i, 7)
+    # (Quarterly) Observations that should evaluate to October of year_string
+    when /^Q(4|04)\b/
+      Date.new(year_string.to_i, 10)
+    else
+     'Error: invalid date %s-%s' % [year_string, other_string]
+    end
```
Comment (vcward) [inline app/lib/data_html_parser.rb]: Rewrote the get_date function with case statement and replaced the arrays with regex. Hopefully this is cleaner 😅. Let me know if you'd like me to make any changes.

Comment (kotodharma) [reply app/lib/data_html_parser.rb]: Erm, well... it's not quite as clean as I would make it :sweat_smile: but the case and regexes are good. I'm approving this, but if you want to see my solution, I couldn't resist coding it up, and you can find it in branch `dave_date_mod`, and take from that anything, or nothing, as you like (note that it is not tested at all). One other thing: the original code just returned nothing on unexpected input, and you've changed that to return an error string, but I'm thinking raising an exception may be better, if we really wanna know where that empty return might have been happening that we'd want to know about. Any questions, lmk.

**Pull request merged**